File: javascript/hamming/hamming.js
var Hamming = function() {};

Hamming.prototype.compute = function(a, b) {
  var hammingDistance = 0;
  var i;
  if (a.length !== b.length) { throw new Error ('DNA strands must be of equal length.'); }
  for (i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) { hammingDistance++; }
  }
  return hammingDistance;
};

module.exports = Hamming;

File: java/hamming/src/main/java/Hamming.java
import javafx.util.Pair;

import java.util.List;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toList;

public class Hamming {
    public static Integer compute(String baseStrand, String compareStrand) {
        if (baseStrand.length() != compareStrand.length()) {
            throw new IllegalArgumentException("Strands are not equal length: base strand is " + baseStrand.length() + ", compare strand is " + compareStrand.length());
        }
        String[] baseStrandArray = baseStrand.split("");
        String[] compareStandArray = compareStrand.split("");
        List<Pair<String, String>> pairs = IntStream.range(0, baseStrand.length()).mapToObj(i ->
            new Pair<>(baseStrandArray[i], compareStandArray[i])
        ).collect(toList());

        return countNonMatchingPairs(pairs);
    }

    private static int countNonMatchingPairs(List<Pair<String, String>> pairs) {
        return (int) pairs.stream().filter(Hamming::doesNotMatch).count();
    }

    private static boolean doesNotMatch(Pair<String, String> pair) {
        return !(pair.getKey().equals(pair.getValue()));
    }
}
